feat: Provisional blocks can be added to archiver

PhilWindle opened the pull request:
This PR makes a number of changes to the archiver.

1. Blocks are now stored separately from checkpoints within the archiver store.
2. Blocks can be added directly to the archiver as well as being synced via checkpoints from L1.
3. Blocks added independently are then deemed to be 'checkpointed' once their checkpoint is received from L1.
4. The archiver now supports multiple blocks per checkpoint.

Some APIs are now considered legacy and will be removed as part of further work. Previously we had `L2Block` and `PublishedL2Block`. We now have `L2BlockNew` and `CheckpointedL2Block`. There are currently APIs for each and this will be updated.

Some operations such manually rolling back blocks still assume 1 block per checkpoint. This will be fixed when we tackle re-orgs of the provisional chain more generally.

spalladino commented on yarn-project/archiver/src/archiver/kv_archiver_store/block_store.ts:
```diff
@@ -83,125 +115,328 @@ export class BlockStore {
-          throw new BlockNumberNotSequentialError(block.block.number, previousBlock.block.number);
+        if (!opts.force && previousBlock && previousBlock.number + 1 !== block.number) {
+          throw new BlockNumberNotSequentialError(block.number, previousBlock.number);
+        }
```
Can we add a similar validation for `indexWithinCheckpoint`?

spalladino replied:
Actually, can we also add validation for `archive` vs `lastArchiveRoot`? Feels like a good check to throw in here (I know it's unrelated to this PR).

PhilWindle replied:
Will do

spalladino commented on yarn-project/archiver/src/archiver/kv_archiver_store/block_store.ts:
```diff
@@ -83,125 +115,328 @@ export class BlockStore {
+        firstBlockCheckpointNumber === INITIAL_CHECKPOINT_NUMBER ||
+        (previousCheckpointNumber !== undefined && previousCheckpointNumber === firstBlockCheckpointNumber - 1);
+      if (!opts.force && !hasPreviousCheckpoint) {
+        throw new InitialCheckpointNumberNotSequentialError(firstBlockCheckpointNumber, previousCheckpointNumber);
```
Just a heads up this will break with build-ahead (not advocating for changing now)

spalladino commented on yarn-project/archiver/src/archiver/kv_archiver_store/block_store.ts:
```diff
@@ -83,125 +115,328 @@ export class BlockStore {
+        for (let i = 0; i < checkpoint.checkpoint.blocks.length; i++) {
+          const block = checkpoint.checkpoint.blocks[i];
+          if (previousBlockNumber !== undefined && previousBlockNumber !== block.number - 1) {
+            throw new BlockNumberNotSequentialError(block.number, previousBlockNumber);
```
We should load the block before the first one to check for `InitialBlockNumberNotSequentialError`

spalladino commented on yarn-project/archiver/src/archiver/kv_archiver_store/block_store.ts:
```diff
@@ -83,125 +115,328 @@ export class BlockStore {
+      }),
+    );
+
+    const converted = await Promise.all(blocksForCheckpoint.map(x => this.getBlockFromBlockStorage(x[0], x[1])));
+    return converted.filter(isDefined);
+  }
+
+  async getProvenBlockNumber(): Promise<BlockNumber> {
+    const provenCheckpointNumber = await this.getProvenCheckpointNumber();
+    const checkpointStorage = await this.#checkpoints.getAsync(provenCheckpointNumber);
+    if (!checkpointStorage) {
+      return BlockNumber(INITIAL_L2_BLOCK_NUM - 1);
+    }
+    return BlockNumber(checkpointStorage.startBlock + checkpointStorage.numBlocks - 1);
+  }
```
We may want to cache this in a future PR, this gets called VERY often from all polling subsystems

spalladino commented on yarn-project/archiver/src/archiver/kv_archiver_store/block_store.ts:
```diff
@@ -340,7 +575,14 @@ export class BlockStore {
+      header,
+      body,
+      CheckpointNumber(blockStorage.checkpointNumber!),
+      0,
```
Heads up this should be the proper index within checkpoint

PhilWindle replied:
Ah, yes good catch!

spalladino commented on yarn-project/archiver/src/archiver/archiver.ts:
```diff
@@ -318,6 +337,51 @@ export class Archiver extends (EventEmitter as new () => ArchiverEmitter) implem
+    // Process each block individually to properly resolve/reject each promise
+    for (const { block, resolve, reject } of queuedItems) {
+      try {
+        await this.store.addBlocks([block]);
```
I understand this'd fail if we try adding a block that had already been synced via L1 (which could happen if a validator is a bit slow). Should we handle that case? No need to do it now though.

PhilWindle replied:
I had planned for that to be handled by the entity pushing the block into the archiver.

spalladino replied:
Can it? Even if the callee checks whether the blocks are there or not, and only pushes them into the queue if they are not in the archiver, we can hit a race condition where the blocks are added via L1 after the caller checked, but before this process method is called.

spalladino commented on yarn-project/archiver/src/archiver/archiver.ts:
```diff
@@ -1331,16 +1360,16 @@ export class Archiver extends (EventEmitter as new () => ArchiverEmitter) implem
    * @returns The requested L2 block.
    */
-  public async getBlock(number: BlockNumber): Promise<L2Block | undefined> {
+  public async getNewBlock(number: BlockNumber): Promise<L2BlockNew | undefined> {
```
Maybe rename to `getL2BlockNew`, so it shows up when we search and replace L2BlockNew to L2Block?

spalladino commented on yarn-project/archiver/src/archiver/archiver.ts:
```diff
@@ -1524,39 +1560,183 @@ export class Archiver extends (EventEmitter as new () => ArchiverEmitter) implem
   }
 
   public async rollbackTo(targetL2BlockNumber: BlockNumber): Promise<void> {
+    // TODO: This still assumes 1 block per checkpoint
```
```suggestion
    // TODO(pw/mbps): This still assumes 1 block per checkpoint
```
Or something so we can search for it later!

spalladino commented on yarn-project/archiver/src/archiver/archiver.ts:
```diff
@@ -1524,39 +1560,183 @@ export class Archiver extends (EventEmitter as new () => ArchiverEmitter) implem
+  }
+
+  public async getBlock(number: BlockNumber): Promise<L2Block | undefined> {
+    // If the number provided is -ve, then return the latest block.
+    if (number < 0) {
+      number = await this.store.getSynchedL2BlockNumber();
+    }
```
We should kill this logic at some point, especially since the `BlockNumber` type checks the value is non negative :-P

spalladino commented on yarn-project/archiver/src/archiver/archiver.ts:
```diff
@@ -1765,59 +1941,113 @@ export class ArchiverStoreHelper
+              this.#updatePublishedContractClasses(contractClassLogs, block.number, Operation.Store),
+              this.#updateDeployedContractInstances(privateLogs, block.number, Operation.Store),
+              this.#updateUpdatedContractInstances(publicLogs, block.header.globalVariables.timestamp, Operation.Store),
+              this.#storeBroadcastedIndividualFunctions(contractClassLogs, block.number),
+            ])
+          ).every(Boolean);
+        }),
```
Maybe we can reuse logic from `addBlocks` above to avoid the duplication?

PhilWindle replied:
Yes, they are the same

spalladino commented on yarn-project/foundation/src/types/index.ts:
```diff
@@ -19,6 +19,11 @@ export function unfreeze<T>(obj: T): Writeable<T> {
   return obj as Writeable<T>;
 }
 
+/** Is defined type guard */
+export function isDefined<T>(value: T | undefined): value is T {
+  return value !== undefined;
+}
```
Nice!

spalladino commented on yarn-project/stdlib/src/block/checkpointed_l2_block.ts:
```diff
@@ -1,18 +1,73 @@
 /**
- * @deprecated `PublishedCheckpoint` is what will be retrieved from L1.
- * L2 blocks do not need to link to L1PublishedData directly.
- * TODO: Create another type (AttestedL2Block?) for an L2 block and its attestations.
+ * Encapsulates an L2 Block along with the checkpoint data associated with it.
  */
+export class CheckpointedL2Block {
```
Good name

spalladino commented on yarn-project/archiver/src/archiver/archiver.test.ts:
I'm amazed you managed to add those new tests with the way this test suite is set up, I found it super difficult to work with.

spalladino reviewed (commented):
Amazing PR. I think the only blocking comment is the hardcoded zero for the index when loading the block from storage. All others are nice-to-have.

spalladino commented on yarn-project/archiver/src/archiver/kv_archiver_store/block_store.ts:
```diff
@@ -209,15 +270,34 @@ export class BlockStore {
+            }
+          } else if (block.indexWithinCheckpoint !== 0) {
+            throw new BlockIndexNotSequentialError(block.indexWithinCheckpoint, undefined);
           }
```
Do we need to also check that block number is 1 here?

PhilWindle replied:
Yes, we probably do

spalladino commented on yarn-project/archiver/src/archiver/archiver.test.ts:
```diff
@@ -114,36 +115,51 @@ describe('Archiver', () => {
-            makeTxOptions: txIndex => ({
-              numPrivateLogs: blockNumber + txIndex,
-            }),
+    // Create checkpoints sequentially to chain archive roots properly.
+    // Each checkpoint's first block's lastArchive must equal the previous block's archive.
```
Ah damn, sorry that the new check required so much additional work on the test suite

spalladino commented on yarn-project/archiver/src/archiver/kv_archiver_store/block_store.ts:
```diff
@@ -371,12 +451,17 @@ export class BlockStore {
 
   async getProvenBlockNumber(): Promise<BlockNumber> {
+    if (this.provenBlockNumber !== undefined) {
+      return this.provenBlockNumber;
+    }
     const provenCheckpointNumber = await this.getProvenCheckpointNumber();
     const checkpointStorage = await this.#checkpoints.getAsync(provenCheckpointNumber);
     if (!checkpointStorage) {
-      return BlockNumber(INITIAL_L2_BLOCK_NUM - 1);
+      this.provenBlockNumber = BlockNumber(INITIAL_L2_BLOCK_NUM - 1);
+    } else {
+      this.provenBlockNumber = BlockNumber(checkpointStorage.startBlock + checkpointStorage.numBlocks - 1);
     }
```
I think we may have a race condition here: the archiver calls `setProvenCheckpointNumber` **before** (and after, but that's not the issue) adding the checkpoints to the db. So if this method gets called inbetween both inserts, we'll incorrectly return `0` as the last proven block number.

PhilWindle replied:
Yes, I have reverted the caching of proven block number. I was concerned about interleaved code paths.

spalladino replied:
I believe we still have the problem outlined above, cache or not. We may need to bring back the old logic to fetch the latest and proven checkpoint numbers, and get the minimum between the two.

PhilWindle replied:
I think that logic still exists. It's within `getProvenCheckpointNumber()`.

spalladino replied:
You're right, my bad!

PhilWindle replied:
Ah no sorry. I get what you mean now and yes I think there is an issue as you say.

spalladino commented:
See https://github.com/AztecProtocol/aztec-packages/pull/19046

PhilWindle committed:
feat: Provisional blocks can be added to archiver

This PR makes a number of changes to the archiver.

1. Blocks are now stored separately from checkpoints within the archiver store.
2. Blocks can be added directly to the archiver as well as being synced via checkpoints from L1.
3. Blocks added independently are then deemed to be 'checkpointed' once their checkpoint is received from L1.
4. The archiver now supports multiple blocks per checkpoint.

Some APIs are now considered legacy and will be removed as part of further work. Previously we had `L2Block` and `PublishedL2Block`. We now have `L2BlockNew` and `CheckpointedL2Block`. There are currently APIs for each and this will be updated.

Some operations such manually rolling back blocks still assume 1 block per checkpoint. This will be fixed when we tackle re-orgs of the provisional chain more generally.

Co-authored-by: Santiago Palladino <santiago@aztec-labs.com>
